### PR TITLE
Added toggleScreen function, using hot key to loop through all screens.

### PR DESCRIPTION
**Pull request by zexee**
Comparing to switchToScreen, it is more handy since the user only need to hit one hotkey.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -32,3 +32,6 @@ CMakeCache.txt
 # Transient in-project-directory dependencies
 /deps/
 /out/build/x64-Debug
+.clangd/
```
Comment (p12tic) [inline .gitignore:35]: Why was this change needed?

File: src/gui/src/Action.cpp [hunk 2]
```diff
@@ -2,11 +2,11 @@
  * barrier -- mouse and keyboard sharing utility
  * Copyright (C) 2012-2016 Symless Ltd.
  * Copyright (C) 2008 Volker Lanz (vl@fidra.de)
- *
```
Comment (p12tic) [inline src/gui/src/Action.cpp:5]: Can we split off the whitespace changes to at least a separate commit?

File: src/gui/src/Action.cpp [hunk 3]
```diff
@@ -45,14 +46,16 @@ Action::Action() :
 
 QString Action::text() const
 {
-    QString text = QString(m_ActionTypeNames[keySequence().isMouseButton() ? type() + 6 : type()  ]) + "(";
+    /* The server config parser don't support () with no argument inside, but support no () */
```
Comment (p12tic) [inline src/gui/src/Action.cpp:49]: I think this comment should be clarified.

File: src/gui/src/Action.h [hunk 4]
```diff
@@ -36,7 +36,7 @@ class Action
     friend QTextStream& operator<<(QTextStream& outStream, const Action& action);
 
     public:
-        enum ActionType { keyDown, keyUp, keystroke, switchToScreen, switchInDirection, lockCursorToScreen, mouseDown, mouseUp, mousebutton };
+        enum ActionType { keyDown, keyUp, keystroke, switchToScreen, toggleScreen, switchInDirection, lockCursorToScreen, mouseDown, mouseUp, mousebutton };
```
Comment (p12tic) [inline src/gui/src/Action.h:39]: I'm wondering if the ActionType enum is converted to an integer in the protocol. Could you please verify which case it is? If we're using integral value of the enum in the protocol, then this new enum should be last.

File: src/lib/server/Config.cpp [hunk 5]
```diff
@@ -1182,6 +1182,10 @@ Config::parseAction(ConfigReadContext& s,
 		action = new InputFilter::SwitchToScreenAction(m_events, screen);
 	}
 
+	else if (name == "toggleScreen") {
```
Comment (p12tic) [inline src/lib/server/Config.cpp:1185]: Can we please use spaces for new code? Different viewers will indent tabs differently.

File: src/lib/server/InputFilter.h [hunk 6]
```diff
@@ -166,7 +166,22 @@ class InputFilter {
...
+
+    // ToggleScreenAction
+    class ToggleScreenAction : public Action {
+    public:
+        ToggleScreenAction(IEventQueue* events);
+
+        // Action overrides
+        virtual Action*        clone() const;
+        virtual String            format() const;
+        virtual void            perform(const Event&);
```
Comment (p12tic) [inline src/lib/server/InputFilter.h:178]: Inconsistent indentation. Could we start the identifier immediately after the type name? This will avoid the need to reindent whenever a long type name is used.

File: src/lib/server/Server.cpp [hunk 7]
```diff
@@ -163,6 +163,10 @@ Server::Server(
 							m_inputFilter,
 							new TMethodEventJob<Server>(this,
 								&Server::handleSwitchToScreenEvent));
+	m_events->adoptHandler(m_events->forServer().toggleScreen(),
```
Comment (p12tic) [inline src/lib/server/Server.cpp:166]: Can we please use spaces for new code? Different viewers will indent tabs differently and this code looks weird e.g. on GitHub.

File: src/lib/server/Server.cpp [hunk 8]
```diff
@@ -1407,6 +1411,22 @@ Server::handleSwitchToScreenEvent(const Event& event, void*)
...
+Server::handleToggleScreenEvent(const Event& event, void*)
+{
+  std::string current = getName(m_active);
+	ClientList::const_iterator index = m_clients.find(current);
+	if (index == m_clients.end()) {
+		LOG((CLOG_DEBUG1 "screen \"%s\" not active", current.c_str()));
+	}
+	else {
+    ++index;
+    if (index == m_clients.end()) index = m_clients.begin();
```
Comment (p12tic) [inline src/lib/server/Server.cpp:1424]: I'd prefer to always use braces for the `if` body. Right now the code is very confusing.

File: .gitignore (hunk 1 above)
Comment (zexee) [reply .gitignore]: removed

File: src/gui/src/Action.cpp (hunk 2 above)
Comment (zexee) [reply src/gui/src/Action.cpp:5]: My vim automatically remove spaces at the end of line. Sorry for that, but I think it is a good behavior.

File: src/gui/src/Action.cpp (hunk 3 above)
Comment (zexee) [reply src/gui/src/Action.cpp]: done

File: src/gui/src/Action.h (hunk 4 above)
Comment (zexee) [reply src/gui/src/Action.h]: in src/gui/src/Action.cpp around my changes.

File: src/lib/server/Config.cpp (hunk 5 above)
Comment (zexee) [reply src/lib/server/Config.cpp:1185]: I don't like tabs either, but this file use tabs. To match the whole file formatting, I use tab here.

File: src/lib/server/InputFilter.h (hunk 6 above)
Comment (zexee) [reply src/lib/server/InputFilter.h:178]: This is also to match the code above and below, to change that, need to change lots of lines, should be in another commit.

File: src/lib/server/Server.cpp (hunk 7 above)
Comment (zexee) [reply src/lib/server/Server.cpp:166]: I don't like tabs either, but this file use tabs. To match the whole file formatting, I use tab here.

File: src/lib/server/Server.cpp (hunk 8 above)
Comment (zexee) [reply src/lib/server/Server.cpp]: done

File: src/gui/src/Action.cpp (hunk 2 above)
Comment (p12tic) [reply src/gui/src/Action.cpp:5]: I agree, just put these changes into a separate commit. Otherwise it will be harder to understand what's happening to whoever to looks into the history later.

File: src/lib/server/InputFilter.h (hunk 6 above)
Comment (p12tic) [reply src/lib/server/InputFilter.h:178]: I don't ask you to change everything, just the new code you're adding. It needs reformatting anyway as it's inconsistent even to itself when viewed through GitHub UI.

File: src/lib/server/Server.cpp (hunk 7 above)
Comment (shymega) [reply src/lib/server/Server.cpp:166]: If the rest of the [Barrier] code uses tabs, we should stick with tabs. Or we can reformat all code to use a unified spacing - either way, we should use one or another to keep it universal.

Comment (p12tic) [reply src/lib/server/Server.cpp:166]: Given that even upstream is switching code bit by bit to spaces, I think we should just use spaces. Tabs are just painful to work with as there's always some tool that does not support it. I worked at a 500 software engineer company that couldn't make tabs work due to that reason and eventually switched everything to spaces.

For us the most sensible solution would be to just use spaces in new code. Converting everything would introduce unnecessary merge conflicts with upstream and will make `git blame` unusable.

Comment (p12tic) [reply src/lib/server/Server.cpp:166]: E.g. 207247b7efe143255d4b57784f67852d55666eb6.

Comment (shymega) [reply src/lib/server/Server.cpp:166]: Alright, let's use spaces in new code then. Ideally converting all of it to spaces would be great, but I think you're right in "keeping a balance" :+!:

Comment (shymega) [reply src/lib/server/Server.cpp:166]: Although I'd be keen to convert tabs to spaces.. I can't see any other way to keep `git blame` working..

Comment (p12tic) [reply src/lib/server/Server.cpp:166]: I think using spaces only for new code shouldn't be a problem. Most editors support setting whitespace settings for modified lines only. The overhead of not committing extra changes accidentally should also be low, because a developer should review his changes before commit anyway.

Comment (shymega) [reply src/lib/server/Server.cpp:166]: Fair enough :smile:

File: src/gui/src/Action.cpp (hunk 2 above)
Comment (zexee) [reply src/gui/src/Action.cpp]: I reverted all space-at-end-of-line change. There are many the case in this code base.

File: src/lib/server/Server.cpp (hunk 7 above)
Comment (zexee) [reply src/lib/server/Server.cpp]: using space now
